KIPs/kip-82.md

```diff
@@ -0,0 +1,376 @@
+
+    # "reward.kip82ratio" parameter (e.g. "20/80") (new)
+    CnMintedBasicRatio: int = 20
+    CnMintedStakeRatio: int = 80
+    CnMintedTotalRatio: int = CnMintedBasicRatio + CnMintedStakeRatio
+
+# CN staking status and KGF/KIR addresses.
+# Corresponds to Klaytn's reward.StakingInfo struct.
+class StakingInfo:
+    KIRAddr: string = ""
+    KGFAddr: string = ""
+    Nodes: List[ConsolidatedNode] = []
+
+# Staking information merged under the same CN.
+# Sometimes a node would register multiple NodeAddrs
+# in which each entry has a different StakingAddr and the same RewardAddr.
+# We treat those entries with common RewardAddr as one node.
+# Corresponds to Klaytn's reward.ConsolidatedNode struct.
+class ConsolidatedNode:
+    NodeAddrs: List[string] = []
+    StakingAddrs: List[string] = []
+    RewardAddr: string = "" # The common reward address of the CN
+    StakingAmount: int = 0  # Total staking amount across StakingAddrs (in KLAY)
+
+# Reward distribution details.
+class RewardSpec:
+    Minted: int = 0   # The amount minted
+    Fee: int = 0      # Total tx fee spent
+    Burnt: int = 0    # The amount burnt
+    Proposer: int = 0    # The amount allocated to the block proposer
+    Stakers: int = 0  # Total amount allocated to stakers
+    Kgf: int = 0      # The amount allocated to KGF
+    Kir: int = 0      # The amount allocated to KIR
+    Rewards: Dict[string, int] = {}  # Mapping from reward recipient to amounts
+
+# Function hierarchy
+# - distribute_block_reward & get_actual_reward
+#   - calc_simple_reward
+#   - calc_deferred_reward
+#     - calc_deferred_fee
+#     - calc_split
+#     - calc_shares
+
+# Distributes a given block's reward at the end of block processing.
+# corresponds to Klaytn's reward.RewardDistributor.DistributeBlockReward() and MintKLAY().
+#
+# - state is the StateDB to apply the rewards.
+# - header is a Header instance.
+# - config is a RewardConfig instance.
+# - staking_info is a StakingInfo instance.
+# - proposerPolicy is the block proposer selection policy.
+#   Valid values are out of RoundRobin, Sticky, WeightedRandom.
+def distribute_block_reward(state, header, config, staking_info, proposerPolicy):
+    if proposerPolicy in [RoundRobin, Sticky]:
+        spec = calc_simple_reward(header, config)
+    else:
+        spec = calc_deferred_reward(header, config, staking_info)
+
+    for addr, amount in spec.Rewards:
+        state.AddBalance(addr, amount)
+
+# Returns the actual reward amounts paid in this block. Used in klay_getReward RPC
+# Returns a RewardSpec.
+def get_actual_reward(header, config, staking_info):
+    if proposerPolicy in [RoundRobin, Sticky]:
+        spec = calc_simple_reward(header, config)
+    else:
+        spec = calc_deferred_reward(header, config, staking_info)
+
+        # If not DeferredTxFee, include block gas fee because they were ignored in
+        # calc_deferred_reward.
+        if not config.DeferredTxFee:
```

Comment:
```suggestion
        if config.DeferredTxFee:
```

If deferred is true, the tx fee calculation is delayed to block reward.

Reply:
Rephrased the comment.

```suggestion
        # Compensate the difference from calc_deferred_reward() and actual payment.
        # If not DeferredTxFee, calc_deferred_reward() assumes 0 total_fee, but
        # nonzero fee should have actually paid to the proposer.
        if not config.DeferredTxFee:
```